aptos-move/framework/aptos-framework/sources/AuctionHouse.move

```diff
@@ -0,0 +1,209 @@
+    const ERROR_AUCTION_NOT_COMPLETE: u64 = 3;
+    const ERROR_NOT_CLAIMABLE: u64 = 4;
+    const ERROR_CLAIM_COINS_FIRST: u64 = 5;
+    const ERROR_ALREADY_CLAIMED: u64 = 6;
+
+    struct AuctionItem has key, store {
+        min_selling_price: u64,
+        duration: u64,
+        start_time: u64,
+        current_bid: u64,
+        current_bidder: address,
+        locked_token: Option<Token>,
+    }
+
+    struct CoinEscrow has key {
+        locked_coins: Table<TokenId, Coin<TestCoin>>,
+    }
+
+    // Set of data sent to the event stream during a auctioning a token
+    struct AuctionEvent has store, drop {
+        id: TokenId,
+        min_selling_price: u64,
+        duration: u64
+    }
+
+    // Set of data sent to the event stream during a bidding for a token
+    struct BidEvent has store, drop {
+        id: TokenId,
+        bid: u64,
+    }
+
+    // Set of data sent to the event stream during a bidding for a token
+    struct ClaimCoinsEvent has store, drop {
+        id: TokenId,
+    }
+
+    // Set of data sent to the event stream during a bidding for a token
+    struct ClaimTokenEvent has store, drop {
+        id: TokenId,
+    }
+
+    struct AuctionData has key {
+        auction_items: Table<TokenId, AuctionItem>,
+        auction_events: EventHandle<AuctionEvent>,
+        bid_events: EventHandle<BidEvent>,
+        claim_coins_events: EventHandle<ClaimCoinsEvent>,
+        claim_token_events: EventHandle<ClaimTokenEvent>,
+    }
+
+    public(script)fun initialize_auction(sender: &signer, creator: address, collection_name: vector<u8>, name: vector<u8>, min_selling_price: u64, duration: u64) acquires AuctionData {
+        let token_id = Token::create_token_id_raw(creator, collection_name, name);
+        let sender_addr = Signer::address_of(sender);
+        if (!exists<AuctionData>(sender_addr)) {
+            move_to(sender, AuctionData {
+                auction_items: Table::new<TokenId, AuctionItem>(),
+                auction_events: Event::new_event_handle<AuctionEvent>(sender),
+                bid_events: Event::new_event_handle<BidEvent>(sender),
+                claim_coins_events: Event::new_event_handle<ClaimCoinsEvent>(sender),
+                claim_token_events: Event::new_event_handle<ClaimTokenEvent>(sender),
+            });
+        };
+        let start_time = Timestamp::now_microseconds();
+        let token = Token::withdraw_token(sender, token_id, 1);
+
+        let auction_data = borrow_global_mut<AuctionData>(sender_addr);
+        let auction_items = &mut auction_data.auction_items;
+
+        // if auction_items still contain token_id, this means that when sender_addr last auctioned this token,
+        // they did not claim the coins from the highest bidder
+        // sender_addr has received the same token somehow but has not claimed the coins from the initial auction
+        assert!(!Table::contains(auction_items, token_id), ERROR_CLAIM_COINS_FIRST);
```

Comment:
or it could mean that the table has an ongoing auction for this token?

I also think we likely will need to rethink this module a little bit to support parallel auctions or other aspects that Bo mentioned ... like offer k and top k bids get them.